### PR TITLE
Replace unsafe_destructor_blind_to_params with may_dangle

### LINKED ISSUE #34761: `#[may_dangle]`, a refined dropck escape hatch (tracking issue for RFC 1327)
Tracking issue for https://github.com/rust-lang/rfcs/pull/1327
- [x] Allow attributes on lifetime/type formal parameters (`generic_param_attrs` feature): https://github.com/rust-lang/rust/pull/34764
- [x] Add `#[may_dangle]` attribute for lifetime/type formal parameters on `unsafe impl<...> Drop` (`dropck_eyepatch` feature)
- [x] Replace uses of `#[unsafe_destructor_blind_to_params]` attribute in libstd with `#[may_dangle]` #38664
- [x] Update nomicon to use `#[may_dangle]` instead of `#[unsafe_destructor_blind_to_params]` #39196
- [x] Remove support for `#[unsafe_destructor_blind_to_params]` attribute (with warning cycle before removal).
  - [x] Deprecated in #38970
  - [x] need to replace remaining references (e.g. in test suite) to `#[unsafe_destructor_blind_to_params]` with `[may_dangle]` before we can actually remove the former.

For docs explaining the feature, also see [the forge](https://forge.rust-lang.org/libs/maintaining-std.html#is-there-a-manual-drop-implementation).

### PR DESCRIPTION
**Pull request by tesuji**
This PR will completely remove support for `#[unsafe_destructor_blind_to_params]` attribute,
which is deprecated in #38970 by `[may_dangle]` unsafe  attribute.

Closes #34761

### CODE REVIEW

**Comment by lzutao** [comment]
Or should I remove support for `#[unsafe_destructor_blind_to_params]` attribute completely?

File: src/test/run-pass/issues/issue-28498-ugeh-ex1.rs [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
+#![feature(dropck_eyepatch)]
 use std::cell::Cell;
 
 struct Concrete<'a>(u32, Cell<Option<&'a Concrete<'a>>>);
 
 struct Foo<T> { data: Vec<T> }
 
-impl<T> Drop for Foo<T> {
+impl<#[may_dangle] T> Drop for Foo<T> {
     // Below is the UGEH attribute
```
Comment (matthewjasper) [inline src/test/run-pass/issues/issue-28498-ugeh-ex1.rs]: Move this comment up a line

**Review by matthewjasper — COMMENTED** [review]
> Or should I remove support for `#[unsafe_destructor_blind_to_params]` attribute completely?

Yes

`#[may_dangle]` is unsafe, so all of the implementations using it need to be `unsafe impl`s.

File: src/test/run-pass/issues/issue-28498-ugeh-ex1.rs (hunk 1 above)
Comment (lzutao) [reply src/test/run-pass/issues/issue-28498-ugeh-ex1.rs]: Yup! I completely forgot about that.

**Comment by lzutao** [comment]
> `#[may_dangle]` is unsafe,

Yup! I completely forgot about that.

**Comment by lzutao** [comment]
The build is passed.

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -199,9 +199,6 @@ declare_features! (
 
     // no-tracking-issue-end
 
-    // Allows using `#[unsafe_destructor_blind_to_params]` (RFC 1238).
-    (active, dropck_parametricity, "1.3.0", Some(28498), None),
```
Comment (matthewjasper) [inline src/libsyntax/feature_gate.rs]: This should be moved to the removed features section lower down in this file

**Review by matthewjasper — COMMENTED** [review]
Could you re-wrap the text in comments that now have a short line?

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -641,6 +638,8 @@ declare_features! (
     (removed, extern_in_paths, "1.33.0", Some(55600), None,
      Some("subsumed by `::foo::bar` paths")),
     (removed, quote, "1.33.0", Some(29601), None, None),
+    // Allows using `#[unsafe_destructor_blind_to_params]` (RFC 1238).
+    (removed, dropck_parametricity, "1.38.0", Some(28498), None, None),
```
Comment (lzutao) [inline src/libsyntax/feature_gate.rs]: Is this correct? This features is unstable since `1.3.0`.

File: src/libsyntax/feature_gate.rs (hunk 2 above)
Comment (lzutao) [reply src/libsyntax/feature_gate.rs]: done.

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (RalfJung) [reply src/libsyntax/feature_gate.rs]: AFAIK unstable features just disappear entirely, no need to keep any trace of them around.

**Pull request merged**